### PR TITLE
kernel: Clarify timeout API regarding negative inputs

### LINKED ISSUE #20438: Kernel timeout API does not document well accepted values
**Describe the bug**
Timeout (and sleep) API has `s32_t` type but it expects that timeout will be non-negative (or K_FOREVER defined as -1). Argument documentation does not specify that which may lead to confusion as user may expect that negative values will be treated as the past and timeout will expire immediately.

Example of that false assumption here: https://github.com/zephyrproject-rtos/zephyr/blob/8e55968abacf5cad6f90c1ceed5a181cf3803bf7/lib/cmsis_rtos_v2/kernel.c#L152

**Expected behavior**
API should clarify that even though arguments are `s32_t` only non-negative values are accepted (or specials).

### PR DESCRIPTION
**Pull request by nordic-krch**
Timeout API use s32_t as an argument but only positive
values are accepted (or special value like K_FOREVER). It was
not specified in the description which may lead to misinterpretation.

Fixes #20438.

### CODE REVIEW

File: kernel/sched.c [hunk 1]
```diff
@@ -975,6 +977,8 @@ s32_t z_impl_k_sleep(int ms)
 {
 	s32_t ticks;
 
+	__ASSERT(ms >= 0, "Only positive values are accepted.");
```
Comment (pabigot) [inline kernel/sched.c]: Can't do this.  Existing code passes `K_FOREVER` which will fail.  That needs to get cleaned up.

Comment (nordic-krch) [reply kernel/sched.c]: in the tree?

Comment (pabigot) [reply kernel/sched.c]: Yes.  Ten examples in-tree, probably more outside.  `k_sleep(K_FOREVER)` is an obvious thing to do, even though it doesn't work as people expect unless it's within a `while (true)` loop.

**Comment by andrewboie** [comment]
> not specified in the description which may lead to misinterpretation.

Misinterpretation?
Users might expect Zephyr to be capable of traveling through time?

**Comment by pabigot** [comment]
> > not specified in the description which may lead to misinterpretation.
> 
> Misinterpretation?
> Users might expect Zephyr to be capable of traveling through time?

Perhaps.  But more likely they might expect code that was working as desired before #19591 would continue to behave the same way in v2.1.

**Comment by nordic-krch** [comment]
@andrewboie 
> Users might expect Zephyr to be capable of traveling through time

That's for sure one thing. Another is that user may expect that he can perform arithmetic operations on timeout or sleep and don't care about negative result which will be interpreted (and currently is, at least until #19591) as the past.

File: kernel/sched.c (hunk 1 above)
Comment (nordic-krch) [reply kernel/sched.c]: moved it to #20448.

**Review by tyrantlizard77 — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@pabigot can you take one more look?

File: kernel/sched.c [hunk 2]
```diff
@@ -368,6 +368,7 @@ static void pend(struct k_thread *thread, _wait_q_t *wait_q, s32_t timeout)
 	}
 
 	if (timeout != K_FOREVER) {
+		__ASSERT(timeout >= 0, "Only positive values are accepted.");
```
Comment (pabigot) [inline kernel/sched.c]: Non-negative, not positive.

Comment (nordic-krch) [reply kernel/sched.c]: done

**Review by pabigot — APPROVED** [review]
OK within the limits of time and release schedule.  It'll have to be updated again if/when we evolve the type and units used to represent timeouts.

**Comment by carlescufi** [comment]
@andrewboie feel free to merge or wait for @andyross 's review

**Comment by andrewboie** [comment]
@carlescufi let me ping @andyross just to be sure

File: kernel/sched.c [hunk 3]
```diff
@@ -368,7 +368,16 @@ static void pend(struct k_thread *thread, _wait_q_t *wait_q, s32_t timeout)
...
 	if (timeout != K_FOREVER) {
-		s32_t ticks = _TICK_ALIGN + k_ms_to_ticks_ceil32(timeout);
+		s32_t ticks;
+
+		__ASSERT(timeout >= 0,
+			"Only non-negative values are accepted.");
+
+		if (timeout < 0) {
+			timeout = 0;
+		}
```
Comment (andyross) [inline kernel/sched.c]: What's the value of having both an assert and a runtime clamp?  I'd leave the former in and take this out.  Note that this code gets changed in the timeout patch anyway, where conversion happens in the initialization of the k_timeout_t before we get here.

**Review by andyross — APPROVED** [review]
Doc changes look great.  Not loving the "assert and clamp" behavior, but it's harmless.  This does complicate the rebase of the timeout patches I guess, but that's not awful.

File: kernel/sched.c (hunk 3 above)
Comment (nashif) [reply kernel/sched.c]: Please removing the the runtime error checking/handling, or remove the assert and handle wrong parameter differently. This is similar to the other PR with k_sleep() and negative values. see https://github.com/zephyrproject-rtos/zephyr/pull/16702 and references in that PR for why.

Comment (nashif) [reply kernel/sched.c]: agree with andyross here, @andrewboie  merged while I was commenting, so we will deal with this later when doing the overall runtime checks

**Pull request merged**